### ISSUE TITLE
MapTest-058

### ISSUE THREAD

**Issue by michaelhkay** [qt4cg/qt4tests#109]
Test case MapTest-058 does

```
let $s as function(xs:anyAtomicType) as xs:string? := map { 1:'A','x':'B' }
            return $s?*
```

and it has recently started failing on Saxon. The reason is https://github.com/qt4cg/qtspecs/pull/1023 , in which I discovered that function coercion needs to happen even in cases where the supplied function is already an instance of the required type, in order that arguments are properly checked.

So Saxon is now doing function coercion in this case, and the result of function coercion is not a map, so the lookup operator is failing.

I think this interpretation is probably correct. Consider the case

```
let $s as function(xs:anyAtomicType) as xs:string? := map { 1:'A','x':'B' }
            return $s(())
```

As a consequence of the reasoning in PR1023, we should reject this on the grounds that `function(xs:anyAtomicType) as xs:string` does not accept () as an argument, even though the map does. So the function coercion does appear to be necessary, and it does have the consequence of producing something that is not a map.

So I think the test is incorrect: once you have coerced a map to a function type, you can't use it as a map. This would be true even if the argument were declared as `xs:anyAtomicType?` - the spec says function coercion happens whether or not it is actually needed, and the consequence is that we end up with a function that isn't a map.

**Comment by michaelhkay** [comment]
I am changing the test to expect a type error. I have raised a PR against the spec to explain the reasoning.

**Comment by ChristianGruen** [comment]
I assume the following expression is still successful?

```xquery
let $s := map { 1: 'A', 'x': 'B' } treat as function(xs:anyAtomicType) as xs:string?
return $s?*
```

**Comment by michaelhkay** [comment]
Yes, I think that expression works. I've added it as MapTest-073.

**Comment by ChristianGruen** [comment]
Resolved (0247233d5b55fac191a0ce4006721df126154f96)
